File: sdk/kanari_pay/lib/src/ui/screens/register_screen.dart
import 'package:flutter/material.dart';

import '../../client/auth_client.dart';
import '../widgets/app_ui.dart';

class KanariRegisterScreen extends StatefulWidget {
  final KanariAuthClient authClient;
  final VoidCallback? onRegistrationSuccess;

  const KanariRegisterScreen({
    super.key,
    required this.authClient,
    this.onRegistrationSuccess,
  });

  @override
  State<KanariRegisterScreen> createState() => _KanariRegisterScreenState();
}

class _KanariRegisterScreenState extends State<KanariRegisterScreen> {
  final _formKey = GlobalKey<FormState>();
  final _emailController = TextEditingController();
  final _passwordController = TextEditingController();
  final _confirmPasswordController = TextEditingController();

  bool _isLoading = false;
  bool _obscurePassword = true;
  bool _obscureConfirmPassword = true;
  String? _errorMessage;
  String _selectedCurveType = 'ed25519';

  final List<Map<String, String>> _curveTypes = [
    {'value': 'ed25519', 'label': 'Ed25519', 'desc': 'Fast and secure default'},
    {
      'value': 'k256',
      'label': 'K256',
      'desc': 'Bitcoin and Ethereum compatible',
    },
    {'value': 'p256', 'label': 'P256', 'desc': 'NIST enterprise standard'},
    {
      'value': 'ed25519dilithium3',
      'label': 'Ed25519 + Dilithium3',
      'desc': 'Hybrid classical and PQ',
    },
    {
      'value': 'k256dilithium3',
      'label': 'K256 + Dilithium3',
      'desc': 'Hybrid EVM and PQ',
    },
  ];

  @override
  void dispose() {
    _emailController.dispose();
    _passwordController.dispose();
    _confirmPasswordController.dispose();
    super.dispose();
  }

  Future<void> _handleRegister() async {
    if (!_formKey.currentState!.validate()) return;

    setState(() {
      _isLoading = true;
      _errorMessage = null;
    });

    try {
      final response = await widget.authClient.register(
        email: _emailController.text.trim(),
        password: _passwordController.text,
        curveType: _selectedCurveType,
      );

      if (!mounted) return;

      setState(() {
        _isLoading = false;
      });

      if (response.success) {
        final walletAddr = response.data?.walletAddress;
        showAppSuccessSnackBar(
          context,
          walletAddr != null
              ? 'Registration successful! Wallet: ${walletAddr.substring(0, 10)}...'
              : 'Registration successful!',
        );
        widget.onRegistrationSuccess?.call();
      } else {
        setState(() {
          _errorMessage = response.error ?? 'Registration failed';
        });
      }
    } catch (e) {
      if (!mounted) return;

      setState(() {
        _isLoading = false;
        _errorMessage = 'Network error: $e';
      });
    }
  }

  String? _validatePassword(String? value) {
    if (value == null || value.isEmpty) {
      return 'Please enter a password';
    }
    if (value.length < 8) {
      return 'Password must be at least 8 characters';
    }
    if (!value.contains(RegExp(r'[A-Z]'))) {
      return 'Password must contain at least one uppercase letter';
    }
    if (!value.contains(RegExp(r'[a-z]'))) {
      return 'Password must contain at least one lowercase letter';
    }
    if (!value.contains(RegExp(r'[0-9]'))) {
      return 'Password must contain at least one digit';
    }
    if (!value.contains(RegExp(r'[!@#$%^&*(),.?":{}|<>]'))) {
      return 'Password must contain at least one special character';
    }
    return null;
  }

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    final colorScheme = theme.colorScheme;
    final selectedCurve = _curveTypes.firstWhere(
      (curve) => curve['value'] == _selectedCurveType,
    );

    return AppGradientScaffold(
      appBar: AppBar(title: const Text('Register'), centerTitle: true),
      body: SingleChildScrollView(
        padding: const EdgeInsets.all(16),
        child: Center(
          child: ConstrainedBox(
            constraints: const BoxConstraints(maxWidth: 520),
            child: Form(
              key: _formKey,
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.stretch,
                children: [
                  const SizedBox(height: 16),
                  const AuthHero(
                    icon: Icons.person_add_alt_1_rounded,
                    title: 'Create Your Account',
                    subtitle:
                        'Set up your Kanari account and choose the wallet cryptography that fits your use case.',
                  ),
                  const SizedBox(height: 16),
                  AppPanel(
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.stretch,
                      children: [
                        const AppSectionTitle('Account Details'),
                        const SizedBox(height: 12),
                        TextFormField(
                          controller: _emailController,
                          keyboardType: TextInputType.emailAddress,
                          enabled: !_isLoading,
                          decoration: const InputDecoration(
                            labelText: 'Email',
                            hintText: 'user@example.com',
                            prefixIcon: Icon(Icons.alternate_email_rounded),
                          ),
                          validator: (value) {
                            if (value == null || value.isEmpty) {
                              return 'Please enter your email';
                            }
                            if (!value.contains('@')) {
                              return 'Please enter a valid email';
                            }
                            return null;
                          },
                        ),
                        const SizedBox(height: 16),
                        TextFormField(
                          controller: _passwordController,
                          obscureText: _obscurePassword,
                          enabled: !_isLoading,
                          onChanged: (_) => setState(() {}),
                          decoration: InputDecoration(
                            labelText: 'Password',
                            hintText: 'Strong password required',
                            prefixIcon: const Icon(Icons.lock_rounded),
                            suffixIcon: IconButton(
                              icon: Icon(
                                _obscurePassword
                                    ? Icons.visibility_rounded
                                    : Icons.visibility_off_rounded,
                              ),
                              onPressed: () {
                                setState(() {
                                  _obscurePassword = !_obscurePassword;
                                });
                              },
                            ),
                          ),
                          validator: _validatePassword,
                        ),
                        const SizedBox(height: 12),
                        _RequirementsCard(password: _passwordController.text),
                        const SizedBox(height: 16),
                        TextFormField(
                          controller: _confirmPasswordController,
                          obscureText: _obscureConfirmPassword,
                          enabled: !_isLoading,
                          decoration: InputDecoration(
                            labelText: 'Confirm Password',
                            hintText: 'Re-enter your password',
                            prefixIcon: const Icon(Icons.verified_user_rounded),
                            suffixIcon: IconButton(
                              icon: Icon(
                                _obscureConfirmPassword
                                    ? Icons.visibility_rounded
                                    : Icons.visibility_off_rounded,
                              ),
                              onPressed: () {
                                setState(() {
                                  _obscureConfirmPassword =
                                      !_obscureConfirmPassword;
                                });
                              },
                            ),
                          ),
                          validator: (value) {
                            if (value == null || value.isEmpty) {
                              return 'Please confirm your password';
                            }
                            if (value != _passwordController.text) {
                              return 'Passwords do not match';
                            }
                            return null;
                          },
                        ),
                      ],
                    ),
                  ),
                  const SizedBox(height: 18),
                  AppPanel(
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.stretch,
                      children: [
                        const AppSectionTitle('Wallet Cryptography'),
                        const SizedBox(height: 8),
                        Text(
                          selectedCurve['desc']!,
                          style: theme.textTheme.bodyMedium?.copyWith(
                            color: colorScheme.onSurfaceVariant,
                          ),
                        ),
                        const SizedBox(height: 16),
                        DropdownButtonFormField<String>(
                          initialValue: _selectedCurveType,
                          decoration: const InputDecoration(
                            labelText: 'Curve Type',
                            prefixIcon: Icon(Icons.security_rounded),
                          ),
                          items: _curveTypes.map((curve) {
                            return DropdownMenuItem<String>(
                              value: curve['value'],
                              child: Text(curve['label']!),
                            );
                          }).toList(),
                          onChanged: _isLoading
                              ? null
                              : (value) {
                                  if (value == null) return;
                                  setState(() {
                                    _selectedCurveType = value;
                                  });
                                },
                        ),
                      ],
                    ),
                  ),
                  if (_errorMessage != null) ...[
                    const SizedBox(height: 16),
                    AppErrorBanner(message: _errorMessage!),
                  ],
                  const SizedBox(height: 24),
                  AppWideButton(
                    onPressed: _isLoading ? null : _handleRegister,
                    icon: Icons.person_add_alt_1_rounded,
                    label: 'Create Account',
                    child: _isLoading
                        ? const SizedBox(
                            width: 20,
                            height: 20,
                            child: CircularProgressIndicator(strokeWidth: 2),
                          )
                        : const Text('Create Account'),
                  ),
                  const SizedBox(height: 12),
                  AppWideButton(
                    onPressed: _isLoading ? null : () => Navigator.pop(context),
                    icon: Icons.login_rounded,
                    label: 'Already have an account? Login',
                    style: AppWideButtonStyle.text,
                  ),
                ],
              ),
            ),
          ),
        ),
      ),
    );
  }
}

class _RequirementsCard extends StatelessWidget {
  final String password;

  const _RequirementsCard({required this.password});

  @override
  Widget build(BuildContext context) {
    final colorScheme = Theme.of(context).colorScheme;

    return Container(
      padding: const EdgeInsets.all(14),
      decoration: BoxDecoration(
        color: colorScheme.surfaceContainerHighest,
        borderRadius: BorderRadius.circular(12),
        border: Border.all(color: colorScheme.outline.withValues(alpha: 0.2)),
      ),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Text(
            'Password requirements',
            style: Theme.of(
              context,
            ).textTheme.labelLarge?.copyWith(fontWeight: FontWeight.w700),
          ),
          const SizedBox(height: 8),
          _RequirementRow(
            text: 'At least 8 characters',
            isValid: RegExp(r'.{8,}').hasMatch(password),
          ),
          _RequirementRow(
            text: 'One uppercase letter',
            isValid: RegExp(r'[A-Z]').hasMatch(password),
          ),
          _RequirementRow(
            text: 'One lowercase letter',
            isValid: RegExp(r'[a-z]').hasMatch(password),
          ),
          _RequirementRow(
            text: 'One digit',
            isValid: RegExp(r'[0-9]').hasMatch(password),
          ),
          _RequirementRow(
            text: 'One special character',
            isValid: RegExp(r'[!@#$%^&*(),.?":{}|<>]').hasMatch(password),
          ),
        ],
      ),
    );
  }
}

class _RequirementRow extends StatelessWidget {
  final String text;
  final bool isValid;

  const _RequirementRow({required this.text, required this.isValid});

  @override
  Widget build(BuildContext context) {
    final colorScheme = Theme.of(context).colorScheme;

    return Padding(
      padding: const EdgeInsets.symmetric(vertical: 3),
      child: Row(
        children: [
          Icon(
            isValid
                ? Icons.check_circle_rounded
                : Icons.radio_button_unchecked_rounded,
            size: 16,
            color: isValid ? Colors.green : colorScheme.onSurfaceVariant,
          ),
          const SizedBox(width: 8),
          Expanded(
            child: Text(
              text,
              style: TextStyle(
                fontSize: 12,
                color: isValid ? Colors.green : colorScheme.onSurfaceVariant,
              ),
            ),
          ),
        ],
      ),
    );
  }
}

File: sdk/kanariapp/app/src/main/java/com/jamesatomc/kanariapp/ui/screens/RegisterScreen.kt
package com.jamesatomc.kanariapp.ui.screens

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.AlternateEmail
import androidx.compose.material.icons.filled.Lock
import androidx.compose.material.icons.filled.PersonAdd
import androidx.compose.material.icons.filled.Visibility
import androidx.compose.material.icons.filled.VisibilityOff
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.Cancel
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.text.input.VisualTransformation
import androidx.compose.ui.unit.dp
import com.jamesatomc.kanariapp.network.KanariClient
import com.jamesatomc.kanariapp.network.models.KanariEnvironment
import com.jamesatomc.kanariapp.network.models.RegisterRequest
import com.jamesatomc.kanariapp.ui.components.AuthHeroSection
import com.jamesatomc.kanariapp.ui.components.ErrorBanner
import com.jamesatomc.kanariapp.ui.components.LoadingButton
import kotlinx.coroutines.launch

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun RegisterScreen(
    onRegisterSuccess: () -> Unit,
    onNavigateToLogin: () -> Unit
) {
    var email by remember { mutableStateOf("") }
    var password by remember { mutableStateOf("") }
    var confirmPassword by remember { mutableStateOf("") }
    var isLoading by remember { mutableStateOf(false) }
    var error by remember { mutableStateOf<String?>(null) }
    var passwordVisible by remember { mutableStateOf(false) }
    var confirmVisible by remember { mutableStateOf(false) }

    val scope = rememberCoroutineScope()
    val client = remember { KanariClient(KanariEnvironment.dev) }

    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text("Register") },
                colors = TopAppBarDefaults.topAppBarColors(containerColor = MaterialTheme.colorScheme.surface)
            )
        },
        containerColor = MaterialTheme.colorScheme.surface
    ) { padding ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(padding)
                .verticalScroll(rememberScrollState())
                .padding(horizontal = 16.dp),
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            Spacer(Modifier.height(16.dp))

            AuthHeroSection(
                icon = Icons.Default.PersonAdd,
                title = "Create Your Account",
                subtitle = "Set up your Kanari account and choose the wallet that fits your use case."
            )

            Spacer(Modifier.height(16.dp))

            Surface(
                modifier = Modifier.fillMaxWidth(),
                shape = RoundedCornerShape(18.dp),
                color = MaterialTheme.colorScheme.surfaceContainerLowest,
                border = androidx.compose.foundation.BorderStroke(1.dp, MaterialTheme.colorScheme.outlineVariant)
            ) {
                Column(modifier = Modifier.padding(16.dp), verticalArrangement = Arrangement.spacedBy(16.dp)) {
                    Text("Account Details", style = MaterialTheme.typography.titleMedium)

                    OutlinedTextField(
                        value = email,
                        onValueChange = { email = it },
                        label = { Text("Email") },
                        placeholder = { Text("user@example.com") },
                        leadingIcon = { Icon(Icons.Default.AlternateEmail, contentDescription = null) },
                        modifier = Modifier.fillMaxWidth(),
                        shape = RoundedCornerShape(12.dp),
                        singleLine = true,
                        keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Email),
                        enabled = !isLoading
                    )

                    OutlinedTextField(
                        value = password, onValueChange = { password = it },
                        label = { Text("Password") }, placeholder = { Text("Strong password required") },
                        leadingIcon = { Icon(Icons.Default.Lock, contentDescription = null) },
                        trailingIcon = {
                            IconButton(onClick = { passwordVisible = !passwordVisible }) {
                                Icon(
                                    if (passwordVisible) Icons.Default.VisibilityOff else Icons.Default.Visibility,
                                    contentDescription = null
                                )
                            }
                        },
                        visualTransformation = if (passwordVisible) VisualTransformation.None else PasswordVisualTransformation(),
                        modifier = Modifier.fillMaxWidth(), shape = RoundedCornerShape(12.dp),
                        singleLine = true, enabled = !isLoading
                    )

                    Surface(
                        modifier = Modifier.fillMaxWidth(), shape = RoundedCornerShape(12.dp),
                        color = MaterialTheme.colorScheme.surfaceContainerHighest,
                        border = androidx.compose.foundation.BorderStroke(
                            1.dp,
                            MaterialTheme.colorScheme.outline.copy(alpha = 0.2f)
                        )
                    ) {
                        Column(modifier = Modifier.padding(14.dp)) {
                            Text("Password requirements", style = MaterialTheme.typography.labelLarge)
                            Spacer(Modifier.height(8.dp))
                            RequirementRow("At least 8 characters", password.length >= 8)
                            RequirementRow("One uppercase letter", password.contains(Regex("[A-Z]")))
                            RequirementRow("One lowercase letter", password.contains(Regex("[a-z]")))
                            RequirementRow("One digit", password.contains(Regex("[0-9]")))
                            RequirementRow(
                                "One special character",
                                password.contains(Regex("[!@#\$%^&*(),.?\":{}|<>]"))
                            )
                        }
                    }

                    OutlinedTextField(
                        value = confirmPassword, onValueChange = { confirmPassword = it },
                        label = { Text("Confirm Password") }, placeholder = { Text("Re-enter your password") },
                        leadingIcon = { Icon(Icons.Default.Lock, contentDescription = null) },
                        trailingIcon = {
                            IconButton(onClick = { confirmVisible = !confirmVisible }) {
                                Icon(
                                    if (confirmVisible) Icons.Default.VisibilityOff else Icons.Default.Visibility,
                                    contentDescription = null
                                )
                            }
                        },
                        visualTransformation = if (confirmVisible) VisualTransformation.None else PasswordVisualTransformation(),
                        modifier = Modifier.fillMaxWidth(), shape = RoundedCornerShape(12.dp),
                        singleLine = true, enabled = !isLoading
                    )
                }
            }

            Spacer(Modifier.height(16.dp))

            Surface(
                modifier = Modifier.fillMaxWidth(), shape = RoundedCornerShape(18.dp),
                color = MaterialTheme.colorScheme.surfaceContainerLowest,
                border = androidx.compose.foundation.BorderStroke(1.dp, MaterialTheme.colorScheme.outlineVariant)
            ) {
                Column(modifier = Modifier.padding(16.dp), verticalArrangement = Arrangement.spacedBy(8.dp)) {
                    Text("Wallet Cryptography", style = MaterialTheme.typography.titleMedium)
                    Text(
                        "Ed25519 — Fast and secure default",
                        style = MaterialTheme.typography.bodyMedium,
                        color = MaterialTheme.colorScheme.onSurfaceVariant
                    )
                }
            }

            Spacer(Modifier.height(16.dp))

            if (error != null) {
                ErrorBanner(error = error!!); Spacer(Modifier.height(16.dp))
            }

            LoadingButton(
                onClick = {
                    scope.launch {
                        isLoading = true; error = null
                        try {
                            val response = client.authService.register(RegisterRequest(email, password))
                            if (response.isSuccessful && response.body()?.success == true) onRegisterSuccess()
                            else error = response.body()?.error ?: "Registration failed"
                        } catch (e: Exception) {
                            error = e.message
                        }
                        isLoading = false
                    }
                },
                text = "Create Account",
                icon = Icons.Default.PersonAdd,
                enabled = email.isNotEmpty() && password.isNotEmpty() && confirmPassword == password,
                isLoading = isLoading,
                modifier = Modifier.fillMaxWidth()
            )

            Spacer(Modifier.height(12.dp))

            TextButton(onClick = onNavigateToLogin, modifier = Modifier.fillMaxWidth().height(56.dp)) {
                Text("Already have an account? Login")
            }

            Spacer(Modifier.height(16.dp))
        }
    }
}

@Composable
private fun RequirementRow(text: String, isValid: Boolean) {
    Row(
        modifier = Modifier.padding(vertical = 3.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        Icon(
            if (isValid) Icons.Default.CheckCircle else Icons.Default.Cancel,
            contentDescription = null, modifier = Modifier.size(16.dp),
            tint = if (isValid) MaterialTheme.colorScheme.tertiary else MaterialTheme.colorScheme.onSurfaceVariant
        )
        Text(
            text, style = MaterialTheme.typography.bodySmall,
            color = if (isValid) MaterialTheme.colorScheme.tertiary else MaterialTheme.colorScheme.onSurfaceVariant
        )
    }
}
